link against jemalloc

not sure if it would make a difference, but I noticed that in bolt's dockerfile they're linking llvm against jemalloc.
https://github.com/llvm/llvm-project/blob/main/bolt/utils/docker/Dockerfile#L23
That might be interesting to try, perhaps if any of the memory allocations are showing up in profiles of kernel builds. I guess we're using whatever malloc implementation musl has? I wonder if it has its own? "mallocng?" https://git.musl-libc.org/cgit/musl/tree/src/malloc/mallocng/malloc.c

Adding the following `-D LLVM_INTEGRATED_CRT_ALLOC=/rpmalloc-1.4.4` to my cmake invocation produces:
```
CMake Error at CMakeLists.txt:575 (message):
  LLVM_INTEGRATED_CRT_ALLOC is only supported on Windows.
```
which is lame. I guess that's the purpose of https://reviews.llvm.org/D101427.

--- a/llvm-project/Dockerfile.epoch3
+++ b/llvm-project/Dockerfile.epoch3
@@ -5,6 +5,7 @@ RUN wget --no-verbose https://git.kernel.org/torvalds/t/linux-5.18-rc6.tar.gz
 RUN wget --no-verbose https://musl.libc.org/releases/musl-1.2.3.tar.gz
 RUN wget --no-verbose https://zlib.net/zlib-1.2.12.tar.gz
 RUN wget --no-verbose https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.1/llvm-project-14.0.1.src.tar.xz
+RUN wget --no-verbose https://github.com/jemalloc/jemalloc/releases/download/5.3.0/jemalloc-5.3.0.tar.bz2
 
 FROM ${BASE} AS stage2
 FROM docker.io/alpine:edge AS stage3
@@ -58,6 +59,19 @@ RUN cd ${ZLIB_DIR} && \
 RUN apk add make
 RUN make -C ${ZLIB_DIR} -j$(nproc)
 RUN make -C ${ZLIB_DIR} -j$(nproc) install
+RUN apk del make
+
+### Jemalloc
+COPY --from=source jemalloc-5.3.0.tar.bz2 .
+RUN tar xf jemalloc-5.3.0.tar.bz2
+ARG JEMALLOC_DIR=jemalloc-5.3.0/build
+RUN mkdir -p ${JEMALLOC_DIR}
+RUN cd ${JEMALLOC_DIR} && \
+  CC=clang AR=llvm-ar NM=llvm-nm CPPFLAGS=${SYSROOT} LDFLAGS=${SYSROOT} \
+  ../configure --disable-libdl --prefix=/usr
+RUN apk add make
+RUN make -C ${JEMALLOC_DIR} -j$(nproc) build_lib_static
+RUN make -C ${JEMALLOC_DIR} -j$(nproc) DESTDIR=/sysroot install_lib_static
 RUN apk del make
 
 ### LLVM

--- a/llvm-project/stage3.cmake
+++ b/llvm-project/stage3.cmake
@@ -13,7 +13,7 @@ set(CMAKE_CXX_FLAGS "--sysroot=/sysroot" CACHE STRING "")
 set(CMAKE_C_FLAGS "--sysroot=/sysroot" CACHE STRING "")
 
 # Statically link resulting executable.
-set(CMAKE_EXE_LINKER_FLAGS "-static -lc++abi" CACHE STRING "")
+set(CMAKE_EXE_LINKER_FLAGS "-static -lc++abi -ljemalloc" CACHE STRING "")
 
 # The compiler builtins are necessary.
 set(COMPILER_RT_BUILD_BUILTINS ON CACHE BOOL "")

